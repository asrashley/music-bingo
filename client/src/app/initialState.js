import { initialState as cards } from '../cards/cardsSlice';
import { initialState as games } from '../games/gamesSlice';
import { initialState as tickets } from '../tickets/ticketsSlice';
import { initialState as user } from '../user/userSlice';

export const initialState = Object.freeze({
    cards,
    games,
    tickets,
    user,
});
