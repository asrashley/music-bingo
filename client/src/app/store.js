import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
