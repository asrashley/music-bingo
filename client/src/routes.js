
export default {
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  register: '/register',
  game: '/game/:gamePk',
  play: '/game/:gamePk/tickets',
  viewTicket: '/game/:gamePk/:ticketPk',
  pastGames: '/history',
  trackListing: '/history/:gamePk',
  listUsers: '/users'
};

