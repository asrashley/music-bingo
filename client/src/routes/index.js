export default {
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  user: '/user',
  changeUser: '/user/modify',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  guestLinks: '/user/guests',
  listUsers: '/user/users',
  guestAccess: '/invite/:token',
  register: '/register',
  listDirectories: '/directories',
  listGames: '/game',
  chooseTickets: '/game/:gameId',
  play: '/game/:gameId/tickets',
  viewTicket: '/game/:gameId/:ticketPk',
  pastGames: '/history',
  trackListing: '/history/:gameId',
  privacy: '/privacy',
};
