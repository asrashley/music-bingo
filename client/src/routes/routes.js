

export const routes = Object.freeze({
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  user: '/user',
  changeUser: '/user/modify',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  guestLinks: '/user/guests',
  settingsIndex: '/user/settings',
  settingsSection: '/user/settings/:section',
  listUsers: '/user/users',
  guestAccess: '/invite/:token',
  register: '/register',
  listDirectories: '/clips',
  listDirectory: '/clips/:dirPk',
  listGames: '/game',
  chooseTickets: '/game/:gameId',
  play: '/game/:gameId/tickets',
  viewTicket: '/game/:gameId/tickets/:ticketPk',
  pastGamesPopularity: '/history',
  pastGamesList: '/history/games',
  pastGamesCalendar: '/history/calendar',
  pastGamesCalendarByYear: '/history/calendar/:year',
  pastGamesByTheme: '/history/themes/:slug',
  gameLastUsed: '/history/themes',
  trackListingByTheme: '/history/themes/:slug/:gameId',
  trackListing: '/history/games/:gameId',
  privacy: '/privacy',
});
