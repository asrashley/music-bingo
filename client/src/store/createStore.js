import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';

import adminReducer from '../admin/adminSlice';
import directoriesReducer from '../directories/directoriesSlice';
import gamesReducer from '../games/gamesSlice';
import messagesReducer from '../messages/messagesSlice';
import settingsReducer from '../settings/settingsSlice';
import systemReducer from '../system/systemSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import usersMiddleware from '../user/userMiddleware';
import messagesMiddleware from '../messages/messagesMiddleware';

import { history } from './history';

export function createStore(preloadedState, logging) {
  const middleware = [...getDefaultMiddleware(), usersMiddleware, messagesMiddleware];
  if (logging === undefined) {
    logging = (process.env.NODE_ENV === 'development');
  }
  if (logging) {
    middleware.push(logger);
  }
  return configureStore({
    middleware,
    reducer: {
      admin: adminReducer,
      directories: directoriesReducer,
      games: gamesReducer,
      messages: messagesReducer,
      settings: settingsReducer,
      system: systemReducer,
      tickets: ticketsReducer,
      router: connectRouter(history),
      user: userReducer,
    },
    preloadedState
  });
}
