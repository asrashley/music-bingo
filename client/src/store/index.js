export { createStore } from './createStore';
export { initialState } from './initialState';
export { history } from './history';
