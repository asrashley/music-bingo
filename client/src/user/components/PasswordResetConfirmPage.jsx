import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { reverse } from 'named-urls';

import { PasswordChangeForm } from './PasswordChangeForm';

import { passwordResetUser } from '../userSlice';
import routes from '../../routes';
import { HistoryPropType } from '../../types/History';

import '../styles/user.scss';

export class PasswordResetConfirmPageComponent extends React.Component {
  static propTypes = {
    dispatch: PropTypes.func,
    history: HistoryPropType.isRequired,
    token: PropTypes.string.isRequired
  };

  state = {
    resetSent: false,
    alert: '',
  };

  onSubmit = (values) => {
    const { history, dispatch } = this.props;
    return dispatch(passwordResetUser(values))
      .then((response) => {
        //console.dir(response);
        if (!response) {
          return {
            type: "validate",
            message: "Unknown error",
            name: "email",
          };
        }
        const { payload } = response;
        if (payload.success === true) {
          history.push(reverse(`${routes.index}`));
          return true;
        }
        this.setState({ alert: payload.error });
        return {
          type: "validate",
          message: payload.error,
          name: "email",
        }
      })
      .catch(err => {
        //console.dir(err);
        const error = (err ? `${err}` : 'Unknown error');
        this.setState({ alert: error });
        return {
          type: "validate",
          message: error,
          name: "email",
        };
      });
  };

  onCancel = (ev) => {
    const { history } = this.props;
    ev.preventDefault();
    history.push(reverse(`${routes.login}`));
    return false;
  };

  render() {
    const { alert } = this.state;
    const { token } = this.props;
    return (
      <div className="reset-sent">
        <PasswordChangeForm alert={alert} onSubmit={this.onSubmit}
          onCancel={this.onCancel} token={token} passwordReset />
      </div>
    );
  }
};

const mapStateToProps = (state, ownProps) => {
  const { token } = ownProps.match.params;
  return {
    token
  }
}

export const PasswordResetConfirmPage = connect(mapStateToProps)(PasswordResetConfirmPageComponent);
