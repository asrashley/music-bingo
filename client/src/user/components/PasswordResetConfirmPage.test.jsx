import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import log from 'loglevel';
import { reverse } from 'named-urls';
import * as reduxReactRouter from '@lagunovsky/redux-react-router';

import { routes } from '../../routes';
import { createStore } from '../../store/createStore';
import { initialState } from '../../store/initialState';
import { fetchMock, renderWithProviders, installFetchMocks, setFormFields } from '../../testHelpers';
import { PasswordResetConfirmPage, PasswordResetConfirmPageComponent } from './PasswordResetConfirmPage';

describe('PasswordResetConfirmPage component', () => {
  let apiMocks;

  beforeEach(() => {
    apiMocks = installFetchMocks(fetchMock, { loggedIn: false });
  });

  afterEach(() => {
    fetchMock.mockReset();
    log.resetLevel();
  });

  it('renders without throwing an exception with initial state', async () => {
    const store = createStore({
      ...initialState,
      routes: {
        params: {
          token: 'token',
        }
      }
    });
    const { findByText, asFragment } = renderWithProviders(
      <PasswordResetConfirmPageComponent dispatch={store.dispatch} token="token" />, { store });
    await findByText(/Confirm New Password/);
    expect(asFragment()).toMatchSnapshot();
  });

  it('redirects if cancel is clicked', async () => {
    const store = createStore({
      ...initialState,
      routes: {
        params: {
          token: 'token',
        }
      }
    });
    const url = await new Promise((resolve) => {
      vi.spyOn(reduxReactRouter, 'push').mockImplementationOnce((url) => {
        resolve(url);
        return {
          type: 'change-location',
          payload: { url },
        };
      });
      renderWithProviders(
        <PasswordResetConfirmPageComponent dispatch={store.dispatch} token="token" />, { store });
      fireEvent.click(screen.getByText("Cancel"));
    });
    expect(url).toEqual(reverse(`${routes.login}`));
  });

  it('sets a new password', async () => {
    const token = 'a.token';
    const store = createStore({
      ...initialState,
      routes: {
        params: {
          token
        }
      }
    });
    const email = 'my.email@address.example';
    const password = 'new.password';
    //log.setLevel('debug');
    const payloadProm = new Promise((resolve) => {
      fetchMock.post('/api/user/reset', (url, opts) => {
        const body = JSON.parse(opts.body);
        resolve(body);
        const { email } = body;
        return apiMocks.jsonResponse({
          email,
          success: true
        });
      });
    });
    const { events, getByText } = renderWithProviders(<PasswordResetConfirmPage dispatch={store.dispatch} />, { store });
    await setFormFields([{
      label: /^Email address/,
      value: email
    }, {
      label: /^New Password/,
      value: password
    }, {
      label: /^Confirm New Password/,
      value: password
    }], events);
    await events.click(getByText("Reset Password"));
    await expect(payloadProm).resolves.toEqual({
      confirmPassword: password,
      email,
      password,
      token
    });
  });

  it('shows an error message if change password fails', async () => {
    const token = 'a.token';
    const store = createStore({
      ...initialState,
      routes: {
        params: {
          token
        }
      }
    });
    const email = 'my.email@address.example';
    const password = 'new.password';
    const error = 'an error message';
    //log.setLevel('debug');
    const payloadProm = new Promise((resolve) => {
      fetchMock.post('/api/user/reset', (url, opts) => {
        const body = JSON.parse(opts.body);
        resolve(body);
        const { email } = body;
        return apiMocks.jsonResponse({
          email,
          error,
          success: false
        });
      });
    });
    const { events } = renderWithProviders(<PasswordResetConfirmPage dispatch={store.dispatch} />, { store });
    await setFormFields([{
      label: /^Email address/,
      value: email
    }, {
      label: /^New Password/,
      value: password
    }, {
      label: /^Confirm New Password/,
      value: password
    }], events);
    await events.click(screen.getByText("Reset Password"));
    await expect(payloadProm).resolves.toEqual({
      confirmPassword: password,
      email,
      password,
      token
    });
    await screen.findByText(error);
  });
});
