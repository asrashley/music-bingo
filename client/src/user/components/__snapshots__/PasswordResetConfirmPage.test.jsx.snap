// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordResetConfirmPage component renders without throwing an exception with initial state 1`] = `
<DocumentFragment>
  <div
    class="reset-sent"
  >
    <form
      class="change-password-form"
    >
      <div
        class="form-group"
      >
        <label
          for="field-email"
        >
          Email address
          <span
            class="required"
          >
            *
          </span>
        </label>
        <input
          class="form-control   "
          id="field-email"
          name="email"
          placeholder="Email address"
          type="text"
        />
        <small
          class="form-text text-muted"
        >
          The email address you used when you registered your account
        </small>
      </div>
      <input
        name="token"
        type="hidden"
        value="token"
      />
      <div
        class="form-group"
      >
        <label
          for="field-password"
        >
          New Password
          <span
            class="required"
          >
            *
          </span>
        </label>
        <input
          class="form-control password  "
          id="field-password"
          name="password"
          placeholder="Choose a new password"
          type="password"
        />
        <small
          class="form-text text-muted"
        >
          The password needs to be at least 5 characters in length.
        </small>
      </div>
      <div
        class="form-group"
      >
        <label
          for="field-confirmPassword"
        >
          Confirm New Password
          <span
            class="required"
          >
            *
          </span>
        </label>
        <input
          class="form-control password  "
          id="field-confirmPassword"
          name="confirmPassword"
          placeholder="Confirm New Password"
          type="password"
        />
        <small
          class="form-text text-muted"
        >
          Please enter the password a second time to make sure you have typed it correctly
        </small>
      </div>
      <div
        class="form-text text-muted"
      >
        * = a required field
      </div>
      <div
        class="form-group modal-footer"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          Reset Password
        </button>
        <button
          class="btn btn-danger"
          name="cancel"
          type="cancel"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
